core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -817,6 +817,9 @@ private final class Sublistener implements ClientStreamListener {
 
     @Override
     public void headersRead(final Metadata headers) {
+      if (substream.previousAttemptCount > 0) {
+        headers.put(GRPC_PREVIOUS_RPC_ATTEMPTS, String.valueOf(substream.previousAttemptCount));
```

Comment:
To avoid confusion, do a `headers.discardAll(GRPC_PREVIOUS_RPC_ATTEMPTS)` just before this.

Reply:
After doing this you can also enhance the test to make sure there is only one value for the key `"grpc-previous-rpc-attempts" . See below